Default IStorage TypeNameHandling.None

EricDahlvang opened the pull request:
Fixes #6013

gabog commented on libraries/Microsoft.Bot.Builder/DictionaryExtensions.cs:
```diff
@@ -0,0 +1,153 @@
+        /// <param name="property">Name of the property.</param>
+        /// <typeparam name="T">The type to which the object should be casted.</typeparam>
+        /// <returns>The in from the dictionary casted to <typeparamref name="T"/>.</returns>
+        public static T MapValueTo<T>(this IDictionary<string, object> dict, string property)
```
Can we rename this method to GetValue<T>?  MapTo is a new term and donet uses GetValue<T> on [JsonNode](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.nodes.jsonnode.getvalue?view=net-6.0) and we also use GetValueAsync<T> in our own `IStatePropertyAccessor`
```suggestion
        public static T GetValue<T>(this IDictionary<string, object> dict, string property)
```

EricDahlvang replied:
This method actually alters the reference and type in the underlying dictionary. It feels like the name should reflect the fact there are side effects.

gabog commented on libraries/Microsoft.Bot.Builder.Dialogs/Prompts/ActivityPrompt.cs:
```diff
@@ -93,7 +88,7 @@ public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc,
 
             // Send initial prompt
-            await OnPromptAsync(dc.Context, (IDictionary<string, object>)state[PersistedState], (PromptOptions)state[PersistedOptions], false, cancellationToken).ConfigureAwait(false);
+            await OnPromptAsync(dc.Context, state.MapValueTo<IDictionary<string, object>>(PersistedState), state.MapValueTo<PromptOptions>(PersistedOptions), false, cancellationToken).ConfigureAwait(false);
```
We probably don't need to call MapTo on PersistedState and PersistedOptions, persistedState is created a few lines above and promptOptions is passed in as an input parameter

gabog commented on libraries/Microsoft.Bot.Builder.Dialogs/Prompts/PromptOptions.cs:
```diff
@@ -51,5 +50,14 @@ public class PromptOptions
+        /// <value>
+        /// Any extra properties to include in the results.
+        /// </value>
+        public IDictionary<string, object> Properties { get; set; } = new Dictionary<string, object>();
```
What is this for? `Properties` doesn't seem to be used anywhere in the solution.

EricDahlvang replied:
Removed

gabog commented on libraries/Microsoft.Bot.Builder/DictionaryExtensions.cs:
```diff
@@ -0,0 +1,153 @@
+        /// <returns>The in from the dictionary casted to <typeparamref name="T"/>.</returns>
+        public static T MapValueTo<T>(this IDictionary<string, object> dict, string property)
+        {
+            if (dict.TryMapValueTo(property, out T result))
```
This method wrapping access to a dictionary, and I think `GetValue<T>` should throw a `KeyNotFoundException` if the required key is not found rather than returning the default to preserve the same behavior. 
Devs should use `TryGetValue<T>` if they are uncertain if the property will be there or not.

gabog commented on libraries/Microsoft.Bot.Builder/IStorage.cs:
```diff
@@ -91,5 +92,25 @@ public static class StorageExtensions
+        /// </summary>
+        /// <param name="value">The object to check for an ETag.</param>
+        /// <returns>The ETag property of the object passed in, if present.</returns>
+        public static string GetETagOrNull(object value)
```
What is this for? Why does it need to be static? seems to be used on BlobStorage and CosmosDbPersistedStorage. Should it be part of the IStorage contract? Can it be internal or private to avoid expanding the surface area?

gabog commented on libraries/Storage/Microsoft.Bot.Builder.Azure.Blobs/BlobsStorage.cs:
```diff
@@ -77,7 +78,7 @@ public BlobsStorage(string dataConnectionString, string containerName, StorageTr
             _jsonSerializer = jsonSerializer ?? JsonSerializer.Create(new JsonSerializerSettings
                                                 {
-                                                    TypeNameHandling = TypeNameHandling.All,
+                                                    TypeNameHandling = TypeNameHandling.None,
```
We probably need update the class and constructor documentation for this class since it is really important that devs know that this will have consequences on how state is managed. We can even tell them that TypeNameHandling.All should be **only** used for backward compat and completely discouraged. Let's chat about how we make this more prominent, we could throw warning to the logger, make the constructor that let's override typehandling obsolete so they get a compiler warning or things like that. 
We should make sure the documentation is consistent among the OOTB IStorage implementations

gabog reviewed (changes requested):
Can we make sure we document these breaking changes in the backward compat doc? When using TypeNameHandling.None, developers will need to use `GetValue<T>` and `TryGetValue<T>` when trying to access properties in `dc.ActiveDialog.State` and `stepContext.Values`
Also, from looking at the source: 
- The `Dialogs.PeristedState` and `Dialogs.PresistedStateKeys` classes don't seem to be used anywhere and can probably be removed.
- `The Dialogs.DialogsInstance.StackIndex` property doesn't seem to be used either

Looking a the serialized data in CosmosDB, there seems to be some inconsistency in the capitalization being used, most properties using camelCase (as it is common for json) but some others like DialogState, StackIndex, Version and ETag, don't. can we be consistent? 

![image](https://user-images.githubusercontent.com/12264946/153893069-bec6bd55-191f-483e-a48c-e30a5f13f6c6.png)

Should we consider make `DialogInstance.State` and WaterfallStepContext.Values strong type state object rather than making them dictionaries? As it is, devs may try to put and get values as if it were dictionaries, that will work within the same turn but not in between turns and they won't realize until runtime. 

Let's chat about ETag offline too, I'd like to get more context on that property.

EricDahlvang commented on libraries/Microsoft.Bot.Builder.Dialogs/Prompts/ActivityPrompt.cs:
```diff
@@ -67,13 +67,8 @@ public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc,
                 throw new ArgumentException($"{nameof(options)} cannot be a cancellation token");
             }
 
-            if (!(options is PromptOptions))
```
Options could be a JObject at this point, until MapValueTo is applied.

EricDahlvang commented:
> Can we make sure we document these breaking changes in the backward compat doc? When using TypeNameHandling.None, developers will need to use `GetValue<T>` and `TryGetValue<T>` when trying to access properties in `dc.ActiveDialog.State` and `stepContext.Values` Also, from looking at the source:
> 
> * The `Dialogs.PeristedState` and `Dialogs.PresistedStateKeys` classes don't seem to be used anywhere and can probably be removed.
> * `The Dialogs.DialogsInstance.StackIndex` property doesn't seem to be used either
> 
> Looking a the serialized data in CosmosDB, there seems to be some inconsistency in the capitalization being used, most properties using camelCase (as it is common for json) but some others like DialogState, StackIndex, Version and ETag, don't. can we be consistent?
> 
> ![image](https://user-images.githubusercontent.com/12264946/153893069-bec6bd55-191f-483e-a48c-e30a5f13f6c6.png)
> 
> Should we consider make `DialogInstance.State` and WaterfallStepContext.Values strong type state object rather than making them dictionaries? As it is, devs may try to put and get values as if it were dictionaries, that will work within the same turn but not in between turns and they won't realize until runtime.
> 
> Let's chat about ETag offline too, I'd like to get more context on that property.

-------------------------------------
- I haven't updated the backward compat doc yet, but certainly will when this is merged.
- Removed PeristedState, PresistedStateKeys and StackIndex
-  DialogState name is based on the name of the type (nameof(DialogState)) so the capitalization is chosen by the compiler. ETag is actually a field generated by cosmosdb, so the capitalization is chosen by cosmosdb.
- Changed `DialogInstance.State` and `WaterfallStepContext.Values` to strongly typed objects CachedBotStateDictionary
- Updated ETag support so it behaves the same across blobs/cosmosdb/memory storages
